### PR TITLE
Boot agama in OSD in s390x with zVM

### PR DESCRIPTION
**Pull request by okynos**
We have added some changes into the code to be able to boot s390x SLE Agama installer.

- Related ticket: https://progress.opensuse.org/issues/167248
- Needles: N/A
- Related PR: https://github.com/jknphy/agama-integration-test-webpack/pull/31
- Verification run: 
  - s390x-zvm SLE: https://openqa.suse.de/tests/15623219
  - s390x-zvm TW: https://openqa.opensuse.org/tests/4543570
  - s390x-zvm Leap: https://openqa.opensuse.org/tests/4543727

### CODE REVIEW

File: tests/yam/agama/agama.pm [hunk 1]
```diff
@@ -25,6 +25,7 @@ sub run {
     my $test = get_required_var('AGAMA_TEST');
     my $reboot_page = $testapi::distri->get_reboot_page();
 
+    sleep(2000);
```
Comment (jknphy) [inline tests/yam/agama/agama.pm:28]: let's drop this.

File: tests/installation/bootloader_s390.pm [hunk 2]
```diff
@@ -330,6 +335,7 @@ sub run {
     # activate console so we can call wait_serial later
     # skip activate serial console since 11sp4 has issue bsc#1159521: iucvconn command not exist
     if (!is_sle('=11-sp4')) {
+        #sleep(2000);
```
Comment (jknphy) [inline tests/installation/bootloader_s390.pm:338]: let's drop this.

File: lib/Distribution/Sle/AgamaDevel.pm [hunk 3]
```diff
@@ -17,12 +17,17 @@ use Yam::Agama::Pom::GrubEntryEditionPage;
...
 sub get_grub_menu {
     return Yam::Agama::Pom::GrubMenuPage->new();
 }
 
 sub get_grub_entry_edition {
-    return Yam::Agama::Pom::GrubEntryEditionPage->new();
+    return is_ppc64le() ? Yam::Agama::Pom::GrubEntryEditionPage->new({
+            number_kernel_line => 3,
+            max_interval => utils::VERY_SLOW_TYPING_SPEED})
+      : Yam::Agama::Pom::GrubEntryEditionPage->new();
```
Comment (jknphy) [inline lib/Distribution/Sle/AgamaDevel.pm:30]: This better in another PR, so this one would be only about s390x, in case we need review from others outside the squad would be more clear.

File: lib/susedistribution.pm [hunk 4]
```diff
@@ -571,10 +571,15 @@ sub init_consoles {
...
-            my $host_name = get_var("AGAMA") ? 'hostname' : 'Hostname';
-            $s390_params .= " $host_ip" . "=${s390_guest_ip}/${s390_guest_subnetmask}";
-            $s390_params .= " $host_name" . "=${s390_guest_hostname}";
+            unless (get_var("AGAMA")){
+                $s390_params .= " HostIP=${s390_guest_ip}/${s390_guest_subnetmask}";
+                $s390_params .= " Hostname" . "=${s390_guest_hostname}";
+            }else{
+                my $s390_gateway_ip = "10.145.10.254";
+                my $s390_domain = "oqa.prg2.suse.org";
+                my $s390_device = "enc800";
```
Comment (jknphy) [inline lib/susedistribution.pm:580]: all these parameter are dynamic depending if runs in O3 or OSD, should be reflected in openQA variable instead of harcoding them.

File: lib/susedistribution.pm [hunk 5]
```diff
@@ -571,10 +571,15 @@ sub init_consoles {
...
             die "Failed to get host by name for '$s390_guest_fqdn' (on " . hostname . ")" unless $packed_ip;
             my $s390_guest_ip = inet_ntoa($packed_ip);
-            my $host_ip = get_var("AGAMA") ? 'hostip' : 'HostIP';
-            my $host_name = get_var("AGAMA") ? 'hostname' : 'Hostname';
-            $s390_params .= " $host_ip" . "=${s390_guest_ip}/${s390_guest_subnetmask}";
-            $s390_params .= " $host_name" . "=${s390_guest_hostname}";
+            unless (get_var("AGAMA")){
+                $s390_params .= " HostIP=${s390_guest_ip}/${s390_guest_subnetmask}";
+                $s390_params .= " Hostname" . "=${s390_guest_hostname}";
+            }else{
```
Comment (jknphy) [inline lib/susedistribution.pm:577]: it is better `if agama` instead of `unless + else`, more readable.

File: lib/susedistribution.pm [hunk 6]
```diff
@@ -571,10 +571,15 @@ sub init_consoles {
...
+                my $s390_gateway_ip = "10.145.10.254";
+                my $s390_domain = "oqa.prg2.suse.org";
```
Comment (jknphy) [inline lib/susedistribution.pm:579]: this could be extracted from the zvm guest right?

File: lib/susedistribution.pm (hunk 4 above)
Comment (jknphy) [reply lib/susedistribution.pm:580]: is it mandatory to add the device?

File: lib/susedistribution.pm [hunk 7]
```diff
@@ -571,10 +571,15 @@ sub init_consoles {
...
             my $s390_guest_ip = inet_ntoa($packed_ip);
-            my $host_ip = get_var("AGAMA") ? 'hostip' : 'HostIP';
-            my $host_name = get_var("AGAMA") ? 'hostname' : 'Hostname';
-            $s390_params .= " $host_ip" . "=${s390_guest_ip}/${s390_guest_subnetmask}";
-            $s390_params .= " $host_name" . "=${s390_guest_hostname}";
+            unless (get_var("AGAMA")){
+                $s390_params .= " HostIP=${s390_guest_ip}/${s390_guest_subnetmask}";
+                $s390_params .= " Hostname" . "=${s390_guest_hostname}";
+            }else{
+                my $s390_gateway_ip = "10.145.10.254";
```
Comment (jknphy) [inline lib/susedistribution.pm:578]: could this be REPO_HOSTO but replacing the last part with 254 ?

File: tests/installation/bootloader_s390.pm [hunk 8]
```diff
@@ -176,18 +178,21 @@ EO_frickin_boot_parms
     ## an empty line and a single "manual=1" at the bottom
     ## of the ftpboot parmfile.  This may fail in obscure
     ## ways when that changes.
-    $s3270->sequence_3270(qw(String(BOTTOM) ENTER String(DELETE) ENTER));
-    $s3270->sequence_3270(qw(String(BOTTOM) ENTER String(DELETE) ENTER));
+    unless (get_var('AGAMA')) {
```
Comment (jknphy) [inline tests/installation/bootloader_s390.pm:181]: why is this needed? seems that this code is there for some sporadic reasons, it works in O3 that I know.

Comment (okynos) [reply tests/installation/bootloader_s390.pm:181]: In Agama the `manual=1` is not present so to avoid last line removal we should not delete the last line of the parmfile.

Comment (jknphy) [reply tests/installation/bootloader_s390.pm:181]: then perhaps a comment would help.

File: tests/installation/bootloader_s390.pm (hunk 2 above)
Comment (okynos) [reply tests/installation/bootloader_s390.pm]: 1825d85eea6b903a2c986748234f860179c5f973

File: lib/Distribution/Sle/AgamaDevel.pm (hunk 3 above)
Comment (okynos) [reply lib/Distribution/Sle/AgamaDevel.pm]: 1825d85eea6b903a2c986748234f860179c5f973

File: lib/susedistribution.pm (hunk 4 above)
Comment (okynos) [reply lib/susedistribution.pm:580]: It is added, take a look at this run https://openqa.opensuse.org/tests/4539131#settings 
`ip=10.150.1.152::10.150.1.254:24:openqanet.opensuse.org:enca00:none`

`enca00` is the device in this case.

Comment (okynos) [reply lib/susedistribution.pm]: 21160037a

File: lib/susedistribution.pm (hunk 5 above)
Comment (okynos) [reply lib/susedistribution.pm]: 21160037a

File: lib/susedistribution.pm (hunk 6 above)
Comment (okynos) [reply lib/susedistribution.pm]: 21160037a

File: lib/susedistribution.pm (hunk 7 above)
Comment (okynos) [reply lib/susedistribution.pm]: 21160037a

File: tests/installation/bootloader_s390.pm (hunk 8 above)
Comment (okynos) [reply tests/installation/bootloader_s390.pm:182]: 21160037a

File: tests/yam/agama/agama.pm [hunk 9]
```diff
@@ -25,6 +25,7 @@ sub run {
     my $test = get_required_var('AGAMA_TEST');
     my $reboot_page = $testapi::distri->get_reboot_page();
 
+    sleep(300);
```
Comment (jknphy) [inline tests/yam/agama/agama.pm:28]: please, do not add sleeps with PR without WIP, they are not allowed in this repo, it is written somewhere about how to contribute (I don't remember where though).

File: tests/yam/agama/patch_agama_tests.pm [hunk 10]
```diff
@@ -9,11 +9,19 @@ use base Yam::Agama::patch_agama_base;
...
 
 sub run {
     select_console 'root-console';
     my ($repo, $branch) = split /#/, get_required_var('YUPDATE_GIT');
     assert_script_run("yupdate patch $repo $branch", timeout => 60);
+
+    if (is_sle() && is_s390x()) {
+        assert_script_run("curl --insecure -o /tmp/ca-certificates-suse.rpm https://download.suse.de/browse/ibs/SUSE:/CA/SLE_15_SP7/noarch/ca-certificates-suse-1.0-150700.8.1.noarch.rpm");
+        assert_script_run("zypper -n install /tmp/ca-certificates-suse.rpm");
+        assert_script_run("systemctl restart agama");
```
Comment (jknphy) [inline tests/yam/agama/patch_agama_tests.pm:23]: this workaround is not related with patching agama with our tests, please add in `agama.pm` with a record_softfailure pointing to a new bug that we should open.

File: lib/susedistribution.pm [hunk 11]
```diff
@@ -571,10 +571,33 @@ sub init_consoles {
...
             my $s390_guest_ip = inet_ntoa($packed_ip);
-            my $host_ip = get_var("AGAMA") ? 'hostip' : 'HostIP';
-            my $host_name = get_var("AGAMA") ? 'hostname' : 'Hostname';
-            $s390_params .= " $host_ip" . "=${s390_guest_ip}/${s390_guest_subnetmask}";
-            $s390_params .= " $host_name" . "=${s390_guest_hostname}";
+            if(get_var("AGAMA")) {
+                my @split = split(/\./, get_required_var("REPO_HOST"));
+                my $s390_gateway_ip = join(".", @split[0], @split[1], @split[2], "254");
+
+                my ($hostname) = split(/\./, get_required_var("ZVM_GUEST"));
```
Comment (jknphy) [inline lib/susedistribution.pm:578]: where do we use this later?

File: lib/susedistribution.pm [hunk 12]
```diff
@@ -571,10 +571,33 @@ sub init_consoles {
...
+
+                    $s390_params .= " ip=${s390_guest_ip}:${s390_guest_hostname}:${s390_gateway_ip}:${s390_guest_subnetmask}:${s390_domain}:${s390_device}:none";
+                }
+                else {
+                    @split = split(/\./, get_required_var("ZVM_GUEST"));
+                    my $s390_domain = join( ".", @split[1], @split[2], @split[3]);
+
+                    my $s390_device = "enca00";
+
+                    $s390_params .= " ip=${s390_guest_ip}:${s390_guest_hostname}:${s390_gateway_ip}:${s390_guest_subnetmask}:${s390_domain}:${s390_device}:none";
```
Comment (jknphy) [inline lib/susedistribution.pm:594]: we are repeating code here, we could just run this once and get the values depending on the case before. Same for other lines that could be similar.

File: lib/susedistribution.pm [hunk 13]
```diff
@@ -571,10 +571,33 @@ sub init_consoles {
...
-            $s390_params .= " $host_name" . "=${s390_guest_hostname}";
+            if(get_var("AGAMA")) {
+                my @split = split(/\./, get_required_var("REPO_HOST"));
+                my $s390_gateway_ip = join(".", @split[0], @split[1], @split[2], "254");
+
+                my ($hostname) = split(/\./, get_required_var("ZVM_GUEST"));
+
+                if(is_sle()){
+                    @split = split(/\./, get_required_var("ZVM_GUEST"));
+                    my $s390_domain = join( ".", @split[1], @split[2], @split[3], @split[4]);
```
Comment (jknphy) [inline lib/susedistribution.pm:582]: the use of index is never a good idea, makes the code harder to maintain, another approach would be to get a list with the values separated by dots and then pop from the list the first element and join the remaining, in that way you can use it for both cases.

File: tests/yam/agama/agama.pm (hunk 9 above)
Comment (jknphy) [inline tests/yam/agama/agama.pm:28]: let's not forget this one after wip is removed ;)

File: lib/susedistribution.pm (hunk 11 above)
Comment (okynos) [reply lib/susedistribution.pm]: 6742aaafb

File: lib/susedistribution.pm (hunk 12 above)
Comment (okynos) [reply lib/susedistribution.pm]: 6742aaafb

File: lib/susedistribution.pm (hunk 13 above)
Comment (okynos) [reply lib/susedistribution.pm]: 6742aaafb

File: tests/yam/agama/patch_agama_tests.pm (hunk 10 above)
Comment (okynos) [reply tests/yam/agama/patch_agama_tests.pm]: 029a14f46

File: tests/yam/agama/agama.pm (hunk 9 above)
Comment (okynos) [reply tests/yam/agama/agama.pm]: e45daadf3

**Comment by okynos** [comment]
Added VR for Leap and TW

In s390x Leap it seems Firefox is not present or not in the same folder.

**Comment by jknphy** [comment]
> Added VR for Leap and TW
> 
> In s390x Leap it seems Firefox is not present or not in the same folder.

Yes I asked for including FF in s390x and ppc64le.
Shouldn't pass the verification when adding the workaround?

**Comment by jknphy** [comment]
LGTM from code perspective, nice job removing indexes btw! only remaining:
- Verification runs updated showing that the workaround make it pass.
- File the bug for the cert and put in the code.
- Squash commits.

**Comment by okynos** [comment]
Do you mean the package installation workaround? Even with this workaround it can't pass green. The DASD devices seems not work as expected, in any case I will rebase all changes on master branch and re-test.

**Comment by jknphy** [comment]
> Do you mean the package installation workaround? Even with this workaround it can't pass green. The DASD devices seems not work as expected, in any case I will rebase all changes on master branch and re-test.

we should not introduce the workaround if we cannot reach the step to run it, it would be better to save it in a branch for next task.
